### PR TITLE
Barnhark/hacks law

### PR DESCRIPTION
**Pull request by kbarnhart**
Addresses #917 

@nicgaspar @nathanlyons @ amforte 

I was going to have to do this anyway to calculate Hacks (I don't have code to do it). 

Let me know what you think.

It relies on the new Channel Profiler (#878 )

I ** think** I"m calculating distance to the divide correctly.

### CODE REVIEW

**Comment by kbarnhart** [comment]
@nathanlyons 

I've started to dig into the distance from divide vs distance from outlet stuff... 

```python
import numpy as np
import matplotlib.pyplot as plt
from landlab import RasterModelGrid, imshow_grid
from landlab.components import FlowAccumulator, FastscapeEroder, ChannelProfiler
from landlab.utils.flow__distance import calculate_flow__distance
from landlab.utils.distance_to_divide import calculate_distance_to_divide


np.random.seed(42)
mg = RasterModelGrid((100, 120), 100)
z = mg.add_zeros("node", "topographic__elevation")
z += np.random.rand(z.size)

fa = FlowAccumulator(mg, flow_director="D8")
sp = FastscapeEroder(mg, K_sp=0.0001, m_sp=0.5, n_sp=1)

dt = 1000
uplift_per_step = 0.001 * dt
core_mask = mg.node_is_core()

for i in range(100):
    z[core_mask] += uplift_per_step
    fa.run_one_step()
    sp.run_one_step(dt=dt)

profiler = ChannelProfiler(mg, number_of_watersheds=1, main_channel_only=True)
profiler.run_one_step()

flowdist = calculate_flow__distance(mg)
dist2divide_long = calculate_distance_to_divide(mg)
dist2divide_short = calculate_distance_to_divide(mg, longest_path=False)

plt.figure()
imshow_grid(mg, dist2divide_long, plot_name="dist2divide_long")
plt.savefig("dist2divide_long.png")
```
dist2divide_long
```python
plt.figure()
imshow_grid(mg, dist2divide_short, plot_name="dist2divide_short")
plt.savefig("dist2divide_short.png")
```
dist2divide_short
```python
plt.figure()
imshow_grid(mg, flowdist, plot_name="flowdist")
plt.savefig("flowdist.png")
```
flowdist
```python
nodes = profiler._profile_structure[0][0]
flow_dist_reverse = max(flowdist[nodes])-flowdist

plt.figure()
imshow_grid(mg, z, cmap="gray", plot_name="dist2divide_long")
plt.scatter(mg.x_of_node[nodes], mg.y_of_node[nodes], c=dist2divide_long[nodes])
plt.colorbar()
plt.savefig("dist2divide_long_single_channel.png")
```
dist2divide_long_single_channel
```python
plt.figure()
imshow_grid(mg, z, cmap="gray", plot_name="flow_dist_reverse")
plt.scatter(mg.x_of_node[nodes], mg.y_of_node[nodes], c=flow_dist_reverse[nodes])
plt.colorbar()
plt.savefig("flow_dist_reverse_single_channel.png")
```
flow_dist_reverse_single_channel

```python
plt.figure()
imshow_grid(mg, z, cmap="gray", plot_name="flow_dist_reverse-dist2divide_long")
plt.scatter(mg.x_of_node[nodes], mg.y_of_node[nodes], c=flow_dist_reverse[nodes]-dist2divide_long[nodes])
plt.colorbar()
plt.savefig("flow_dist_reverse-dist2divide_long_single_channel.png")
```
flow_dist_reverse-dist2divide_long_single_channel

```python
plt.figure()
profiler.plot_profiles(flow_dist_reverse-dist2divide_long)
```
flow_dist_reverse-dist2divide_long_profile


In summary, the distance to the divide utility is off by a constant -141 (length of a diagonal) almost everywhere except the outlet where it is off by 41 (length of a diagonal - length of link). 

So the difference is likely in how the very first and very last distances are calculated.

**Comment by kbarnhart** [comment]
@nathanlyons I figured this out. changes/examples forthcoming

**Comment by kbarnhart** [comment]
@nathanlyons basically there were two issues:
1) The channel profiler wasn't ever able to get ALL the way to the divide due to a drainage threshold issue (this is now fixed over at #878). 
2) My distance_to_divide utility didn't increment the channel length when going from a core node to a boundary nodes. 

Both of these are fixed, and now I find that the flow distance and the distance to divide (long) are inverse along the main channel.

**Comment by nicgaspar** [comment]
Wow @kbarnhart! I'm just starting to look at this, but your example is so beautifully clear and illustrative that I needed to say bravo first. Peeking at the code now.

**Comment by nicgaspar** [comment]
Is it too annoying to change `calculate_hack_coefficients` to `calculate_hack_parameters`?

Just thinking that technically one is a coefficient and one is an exponent. Maybe people would be confused, maybe not.

**Comment by kbarnhart** [comment]
As the Channel Profiler is not py27 compatible, I changed the merge base to `v2_dev`. I've now changed this component to merge into `v2_dev`

File: environment-dev.yml [hunk 1]
```diff
@@ -2,19 +2,9 @@ name: landlab_dev
 channels:
   - conda-forge
 dependencies:
```
Comment (kbarnhart) [inline environment-dev.yml]: @mcflugen These requirements are all listed in `setup.py` am I correct to thing that they only should be listed there and not here as well?

**Comment by kbarnhart** [comment]
Alrighty. I think this is now ready for review. 

@nicgaspar I made your recommended change to turn the term "coefficients" into "parameters". 
@mcflugen I made a few changes to the environment .yml files and setup.py (as this now requires pandas as a dependency). 

@nathanlyons you may also be appropriate to review. 

Note that the PR brings this branch into v2_dev not master.

**Comment by kbarnhart** [comment]
No one seems to have comments on this (beyond the comments from before) so I am merging.

**Pull request merged**